aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+        return transpose ? output.transpose_(-2, -1).contiguous() : output;
+      } else {
+        return at::_unsafe_view(t1_folded.mv(*t2), output_shape);
+      }
+    } else {
+      // Resize output into the correct shape
+      const auto transpose_out = transpose && t2_is_matrix;
+      if (transpose_out) {
+        // Swap last two elements of output_shape
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+        at::native::resize_output(out, output_shape);
+        out.transpose_(-2, -1);
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+      } else {
+        at::native::resize_output(out, output_shape);
+      }
 
-      Tensor res = res_T.reshape(shape).contiguous();
-      return has_out ? out.set_(res) : res;
+      // We then reshape the output to the expected shape and call mm/mv
+      // and transpose back if necessary
+      auto reshaped_out = t2_is_matrix ? out.reshape({batch_product, t2->sizes().back()})
+                                       : out.reshape({batch_product});
+      if (t2_is_matrix) {
+        at::mm_out(reshaped_out, t1_folded, *t2);
+      } else {
+        at::mv_out(reshaped_out, t1_folded, *t2);
+      }
+      if (!reshaped_out.is_alias_of(out)) {
+        out.copy_(reshaped_out.view_as(out));
+      }
+      if (transpose_out) {
+        out.transpose_(-2, -1);
+      }
+      return out;
     }
-  } else if ((dim_tensor1 >= 1 && dim_tensor2 >= 1) && (dim_tensor1 >= 3 || dim_tensor2 >= 3)) {
-    // We are multiplying b1 x n x m1 by x2 x m2 x p (where b1 can be a list);
+  } else {
+    // dim_tensor1 >= 3 || dim_tensor2 >= 3
+    // We are multiplying b1 x n x m1 by b2 x m2 x p (where b1 and b2 can be lists);
     // we track m1 vs m2 separately even though they must match for nicer error messages
-    int64_t n = dim_tensor1 > 1 ? tensor1.size(-2) : 1;
-    int64_t m1 = tensor1.size(-1);
-    IntArrayRef batch_tensor1(tensor1.sizes().data(), std::max<int64_t>(dim_tensor1 - 2, 0));
-    int64_t m2 = dim_tensor2 > 1 ? tensor2.size(-2) : 1;
-    int64_t p = tensor2.size(-1);
-    IntArrayRef batch_tensor2(tensor2.sizes().data(), std::max<int64_t>(dim_tensor2 - 2, 0));
-
-    // expand the batch portion (i.e. cut off matrix dimensions and expand rest)
-    DimVector expand_batch_portion = infer_size_dimvector(batch_tensor1, batch_tensor2);
-
-    DimVector tensor1_expand_size(expand_batch_portion);
-    tensor1_expand_size.push_back(n);
-    tensor1_expand_size.push_back(m1);
-
-    DimVector tensor2_expand_size(expand_batch_portion);
-    tensor2_expand_size.push_back(m2);
-    tensor2_expand_size.push_back(p);
-
-    const int64_t expand_batch_product =
-        c10::multiply_integers(expand_batch_portion);
-
-    std::array<int64_t, 3> tensor1_bmm_view = {expand_batch_product, n, m1};
-
-    std::array<int64_t, 3> tensor2_bmm_view = {expand_batch_product, m2, p};
+    const int64_t n = dim_tensor1 > 1 ? tensor1.sizes().cend()[-2] : 1LL;
```

Comment:
I must admit, what's the point to having `tensor1.size(-2)` if you can't use it in code like here. Can we modify the definition of `size(...)` so that it inlines into the same code here? You basically want a non-bounds-checked version of sizes...

Reply:
cc @swolchok

Reply:
By the way I don't think cend() is saving you an arithmetic op because ArrayRef only stores pointer and length, so to calc cend you still end up doing pointer + length - 2

Reply:
I just use `cend` to be able to use negative indices really.

Fwiw, I believe the problem with `size()` is that it is a virtual function. I don't know how possible would it be to make it non-virtual, but it'd certainly make all this code more readable.

Reply:
we should be able to devirtualize it. If you are interested in continuing to do microoptimizations like this let's talk about what this would entail.

Reply:
I wouldn't mind at all, but I think for now the priority is in working with @mruberry on primtorch, so this will have to wait. Now, it looks like the kind of task that @peterbell10 would like. Otherwise I can look into it in the future.

Reply:
See #75416